pkg/scope/health/health.go

```diff
@@ -0,0 +1,150 @@
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	apps "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/oam-kubernetes-runtime/apis/core/v1alpha2"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/pkg/errors"
+)
+
+const (
+	errNoWorkload            = "could not retrieve workload %q"
+	errNoWorkloadResources   = "could not retrieve resources for workload %q"
+	errResourceNotFound      = "could not retrieve resource %q %q %q"
+	errDeploymentUnavailable = "no ready instance found in %q %q %q"
+
+	defaultTimeout = 10 * time.Second
+)
+
+// UpdateHealthStatus updates the status of the healthscope based on workload resources.
+func UpdateHealthStatus(ctx context.Context, log logging.Logger, client client.Client, healthScope *v1alpha2.HealthScope) error {
+	timeout := defaultTimeout
+	if healthScope.Spec.ProbeTimeout != nil {
+		timeout = time.Duration(*healthScope.Spec.ProbeTimeout) * time.Second
+	}
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	resourceRefs := []runtimev1alpha1.TypedReference{}
+	for _, workloadRef := range healthScope.Spec.WorkloadReferences {
+		// Get workload object.
+		workloadObject := unstructured.Unstructured{}
+		workloadObject.SetAPIVersion(workloadRef.APIVersion)
+		workloadObject.SetKind(workloadRef.Kind)
+		workloadObjectRef := types.NamespacedName{Namespace: healthScope.GetNamespace(), Name: workloadRef.Name}
+		if err := client.Get(ctxWithTimeout, workloadObjectRef, &workloadObject); err != nil {
+			return errors.Wrapf(err, errNoWorkload, workloadRef.Name)
+		}
+
+		// TODO(asouza): Need to find a way to make sure that this is compatible with the scope structure.
+		if value, err := fieldpath.Pave(workloadObject.UnstructuredContent()).GetValue("status.resources"); err == nil {
+			refs := value.([]interface{})
+			for _, item := range refs {
+				ref := item.(map[string]interface{})
+				resourceRef := runtimev1alpha1.TypedReference{
+					APIVersion: fmt.Sprintf("%v", ref["apiVersion"]),
+					Kind:       fmt.Sprintf("%v", ref["kind"]),
+					Name:       fmt.Sprintf("%v", ref["name"]),
+				}
+
+				resourceRefs = append(resourceRefs, resourceRef)
+			}
+		} else {
+			return errors.Wrapf(err, errNoWorkloadResources, workloadRef.Name)
+		}
+	}
+
+	statusc := resourcesHealthStatus(ctxWithTimeout, log, client, healthScope.Namespace, resourceRefs)
+	status := true
+	for r := range statusc {
+		status = status && r
+	}
+
+	health := "unhealthy"
+	if status {
+		health = "healthy"
+	}
+
+	healthScope.Status.Health = health
+	return nil
+}
+
+func resourcesHealthStatus(ctx context.Context, log logging.Logger, client client.Client, namespace string, refs []runtimev1alpha1.TypedReference) <-chan bool {
```

Comment:
nit, plz consider rename "resourcesHealthStatus" and "resourceHealthStatus", with something representing an action and show a bit more differentiation.